### DISCUSSION TITLE
Floating Point data from the tomograms

### DISCUSSION THREAD

**Discussion by sgvdan** [marksgraham/OCT-Converter#39] [General]
First of all, thank you for this incredibly useful contribution!

Going in to how the e2e tomogram data is parsed, the [bitbucket explanation](https://bitbucket.org/uocte/uocte/wiki/Heidelberg%20File%20Format), talks about the data being an unsigned float with 6 bits for exponent and 10 bits for the mantissa.
However, the code here parses it as a 16-bit unsigned-int and then performs a somewhat weird mathematical manipulation to it -
![image](https://user-images.githubusercontent.com/86140273/134891890-b0b13b52-c820-408a-b73a-a1f9458226c2.png)

From my small side-note calculation, this manipulation doesn't coincide with float structure -
![image](https://user-images.githubusercontent.com/86140273/134892577-cefe70a1-76ea-4beb-b93c-3908627ea68a.png)

Would be grateful for some clarification here :)

Thanks!
Dan

**Comment by Dbrown411** [comment]
Hi Dan, 

You'll want to take a look at the look up table LUT. The bytes are first parsed as unsigned integers (for speed), and then converted to the nonstandard float type via the LUT. 

The calculations in the code snippet you posted are for applying gamma to the intensities. The 1/2.4 gamma is pretty arbitrary as far as we can tell.

Dillon

**Reply by sgvdan** [reply]
I did elegantly ignore that part. Will have a look. Thanks!

**Comment by marksgraham** [comment]
Yes thats right - I had initially implemented the conversion this way https://github.com/marksgraham/OCT-Converter/blob/3063aa8e310fdd30d3d4096a61df1bb7a24c781d/oct_converter/readers/e2e.py#L264 and Dillon's implementation gives the same results + a huge speed up

**Reply by sgvdan** [reply]
Indeed a very neat thing to do. Well done!

There is one thing I'm still wondering about: In the `read_custom_float` implementation, you've subtracted 63 from the exponent. However, according to [this](https://en.wikipedia.org/wiki/Exponent_bias) page you are supposed to subtract only ![equation](http://www.sciweavers.org/tex2img.php?eq=2%5E%7B%286-1%29%7D-1%3D32-1%3D31&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=0). Anything I'm missing here?

**Reply by marksgraham** [reply]
I lifted the -63 directly from the uocte [implementation](https://bitbucket.org/uocte/uocte/src/8ed70b7ecd5c92cbbfaac281fe090cf7136cba6d/src/io/load_heidelberg.cpp#lines-63) but I don't know where they got it from.

**Reply by Dbrown411** [reply]
I don't have time to look into this more for a while, and I have a pretty limited background in floating point. But it's my understanding that the typical bias (2^(6-1) - 1) allows for a range of ~(2^-31 - 2^31). So subtracting 63 instead should essentially normalize the pixel intensities to between 0 and 1 (2^-63 - 2^0)? 

Not sure what advantages/effects this has on the distribution of pixel intensities (if any). But maybe this has something to do with the seemingly arbitrary 1/2.4 gamma transformation?

**Reply by sgvdan** [reply]
It basically means that the values we get are all multiplied by 2^(-32).
It is merely a scale factor - that's why I think we haven't noticed it yet. Therefore not a big deal, but indeed a bit weird.
